alexandria/skip_graph.py

```diff
@@ -0,0 +1,303 @@
+        if left is None and right is not None:
+            assert right.get_left_neighbor(level + 1) is None
+        if right is None and left is not None:
+            assert left.get_right_neighbor(level + 1) is None
+        if left is not None and right is not None:
+            assert left.get_right_neighbor(level + 1) == right.key
+            assert right.get_left_neighbor(level + 1) == left.key
+
+        return self._insert_at_level(node, left, right, level + 1)
+
+    def _link_nodes(self,
+                    left: Optional[SGNode],
+                    right: Optional[SGNode],
+                    level: int,
+                    ) -> None:
+        if left is None and right is None:
+            raise Exception("Invariant")
+        elif left is None:
+            right.set_left_neighbor(level, None)
+        elif right is None:
+            left.set_right_neighbor(level, None)
+        else:
+            left.set_right_neighbor(level, right.key)
+            right.set_left_neighbor(level, left.key)
+
+    def _search_insert_point(self,
+                             key: Key,
+                             current: SGNode,
+                             level: int) -> Tuple[Optional[SGNode], Optional[SGNode]]:
+        if key == current.key:
+            raise AlreadyPresent
+        elif key > current.key:
+            for at_level, right_key in current.iter_down_right_levels(level):
+                if right_key is None:
+                    continue
+
+                if key >= right_key:
+                    right_neighbor = self.nodes[right_key]
+                    return self._search_insert_point(key, right_neighbor, at_level)
+            else:
+                right_neighbor_key = current.get_right_neighbor(0)
+                if right_neighbor_key is None:
+                    return current, None
+                right_neighbor = self.nodes[right_neighbor_key]
+                return current, right_neighbor
+        elif key < current.key:
+            for at_level, left_key in current.iter_down_left_levels(level):
+                if left_key is None:
+                    continue
+
+                if key <= left_key:
+                    left_neighbor = self.nodes[left_key]
+                    return self._search_insert_point(key, left_neighbor, at_level)
+            else:
+                left_neighbor_key = current.get_left_neighbor(0)
+                if left_neighbor_key is None:
+                    return None, current
+                left_neighbor = self.nodes[left_neighbor_key]
+                return left_neighbor, current
+        else:
+            raise Exception("Invariant")
+
+    def delete(self, key: Key, current: SGNode) -> None:
+        self.logger.debug('Deleting: %d', key)
+        node = self.search(key, current)
+
+        left: Optional[SGNode]
+        right: Optional[SGNode]
+
+        for level in range(node.max_level, -1, -1):
+            left_neighbor_key = node.get_left_neighbor(level)
```

Comment:
This pattern shows up pretty often! A lot of these methods would get shorter if there was a method like `Graph.left_neighbor(self, node, level) -> Optional[SGNode]`.

Reply:
Yup, this made for easier to read code.